### PR TITLE
Feature: fast progress latents

### PR DESCRIPTION
**Pull request by damian0815**
This pull request adds fast progress images to the web UI by approximating directly from the latent space without passing through the slow VAE+upscale path (thanks to work done elsewhere by @keturn). This enables approximate progress images to be displayed for every diffusion step with effectively no performance penalty.

On the frontend users now have a choice between "Fast" and "Accurate" progress images (or "None"), with "Accurate" preserving the current behaviour (full-res + saving to disk, every 5th step).

Currently defaults to "None" but perhaps it should default to "Fast"? My intuition is that showing the latents like this might help users understand what SD is doing under the hood, especially for img2img or inpainting.

### CODE REVIEW

**Comment by keturn** [comment]
hotgrits did some further analysis on the latent channel to RGB conversion, and came up with this matrix:

> Alternate matrix for converting Stable Diffusion 1.4 latents to RGB + _T*_ with _T*_ being just an extra channel that seems to contain texture information. […] With this you can convert to RGB + Texture and back if you make the inverse matrix via `torch.linalg.inv(matrix)`, unlike the 4 channels to RGB matrix.

```py
[[ 0.3444,  0.1385,  0.0670,  0.4838],
 [ 0.1247,  0.4027,  0.1494, -0.3709],
 [-0.3192,  0.2513,  0.2103,  0.1221],
 [-0.1307, -0.1874, -0.7445,  0.1484]]
```

Even if you don't need the invertible property and just discard that _T*_ channel, I think it does a bit better on colors in general.

These seem to work equally well on SD 1.5. I suspect that's because all these 1.x models are continuations of the original 1.2 model, so the space hasn't changed that much.

**Review by lstein — APPROVED** [review]
Let me know if you think this should be picked into the release candidate. There are a lot of commits here, so it would be good to rebase if you think so.

Probably not till 2.2, what do you think?

**Comment by damian0815** [comment]
i think i'd want @blessedcoolant to approve, as it's making a lot of data throughput when the steps happen quickly - i noticed on a 3090 box today that the web UI would often lag behind the generation

File: backend/invoke_ai_web_server.py [hunk 1]
```diff
@@ -637,6 +636,25 @@ def image_progress(sample, step):
                             "height": height,
                         },
                     )
+
+                if generation_parameters['progress_latents']:
+                    image = self.generate.sample_to_lowres_estimated_image(sample)
+                    (width, height) = image.size
+                    buffered = io.BytesIO()
+                    image.save(buffered, format="PNG")
+                    img_base64 = "data:image/jpeg;base64," + base64.b64encode(buffered.getvalue()).decode('UTF-8')
```
Comment (keturn) [inline backend/invoke_ai_web_server.py:645]: this has `format="PNG"` but then it's given the `image/jpeg` mime type.

**Comment by keturn** [comment]
How's it transmitted? oh, base64 PNG?

I guess if you're talking to the server on localhost, I'd expect it to be able to keep up with that regardless, but yeah. Certainly a lot of room to streamline that.

File: backend/invoke_ai_web_server.py (hunk 1 above)
Comment (damian0815) [reply backend/invoke_ai_web_server.py:645]: so it doooes. spot the non-web programmer (me)

**Comment by damian0815** [comment]
> @torridgristle did some further analysis on the latent channel to RGB conversion, and came up with this matrix:

can you post an original link to that? i'd like to attribute it properly if possible

**Comment by keturn** [comment]
> They can just put my handle @torridgristle as the attribution, but other than that it's just [a message on Discord](https://discord.com/channels/729741769192767510/730484623028519072/1023283006418464898).

**Pull request merged**